outlines/generate/generator.py

```diff
@@ -0,0 +1,362 @@
+import dataclasses
+import math
+from typing import TYPE_CHECKING, Callable, Iterator, List, Optional, Tuple, Union
+
+import torch
+
+from outlines.fsm.fsm import FSMState
+
+if TYPE_CHECKING:
+    from outlines.fsm.fsm import FSM
+    from outlines.generate.samplers import Sampler
+
+
+@dataclasses.dataclass(frozen=True)
+class GenerationState:
+    token_ids: torch.Tensor
+    kv_cache: torch.Tensor
+    logits: torch.Tensor
+
+
+class SequenceGenerator:
+    def __init__(self, fsm, model, sampler, device):
+        self.generate_token = token_generator(model, sampler)
+        self.fsm = fsm
+        self.tokenizer = model.tokenizer
+        self.device = device
+
+    def init_generation_state(
```

Comment:
This should be a function and initiate the FSM states as well so `sequence_generator` can be used independently by advanced users. This class would then belong to `api.py`.